libs/sdk-core/src/error.rs

```diff
@@ -1,60 +1,311 @@
-    PersistenceFailure { err: String },
+    pub fn swap_provider_error(err: String) -> Self {
+        Self::onchain_error(OnchainErrorCode::SwapProviderError, err)
+    }
 
-    #[error("Failed to receive payment: {err}")]
-    ReceivePaymentFailed { err: String },
+    pub fn decoding_failed(err: String) -> Self {
+        Self::value_error(ValueErrorCode::DecodingFailed, err)
+    }
 
-    #[error("Failed to send payment: {err}")]
-    SendPaymentFailed { err: String },
+    pub fn invalid_url(e: String) -> Self {
+        Self::value_error(ValueErrorCode::InvalidUrl, format!("Invalid Url: {e}"))
+    }
 }
 
 impl From<rusqlite::Error> for SdkError {
     fn from(value: rusqlite::Error) -> Self {
-        Self::PersistenceFailure {
-            err: value.to_string(),
-        }
+        Self::persistence_failed(value.to_string())
     }
 }
 
 impl From<rusqlite_migration::Error> for SdkError {
     fn from(value: rusqlite_migration::Error) -> Self {
-        Self::PersistenceFailure {
-            err: value.to_string(),
-        }
+        Self::persistence_failed(value.to_string())
+    }
+}
+
+impl From<serde_json::Error> for SdkError {
+    fn from(value: serde_json::Error) -> Self {
+        Self::decoding_failed(value.to_string())
+    }
+}
+
+impl From<bitcoin::hashes::hex::Error> for SdkError {
+    fn from(value: bitcoin::hashes::hex::Error) -> Self {
+        Self::decoding_failed(value.to_string())
+    }
+}
+
+impl From<bitcoin::util::address::Error> for SdkError {
+    fn from(value: bitcoin::util::address::Error) -> Self {
+        Self::decoding_failed(value.to_string())
+    }
+}
+
+impl From<bitcoin::util::sighash::Error> for SdkError {
+    fn from(value: bitcoin::util::sighash::Error) -> Self {
+        Self::signature_error(value.to_string())
+    }
+}
+
+impl From<hex::FromHexError> for SdkError {
+    fn from(value: hex::FromHexError) -> Self {
+        Self::decoding_failed(value.to_string())
+    }
+}
+
+impl From<reqwest::Error> for SdkError {
+    fn from(value: reqwest::Error) -> Self {
+        Self::request_failed(value.to_string())
+    }
+}
+
+impl From<secp256k1::Error> for SdkError {
+    fn from(value: secp256k1::Error) -> Self {
+        Self::signature_error(value.to_string())
+    }
+}
+
+impl From<tonic::transport::Error> for SdkError {
+    fn from(value: tonic::transport::Error) -> Self {
+        Self::signature_error(value.to_string())
+    }
+}
+
+impl From<tonic::Status> for SdkError {
+    fn from(value: tonic::Status) -> Self {
+        Self::service_error(ServiceErrorCode::TransportError, value.to_string())
     }
 }
 
 // TODO This won't be necessary when all service methods return SdkResult
 impl From<anyhow::Error> for SdkError {
     fn from(value: anyhow::Error) -> Self {
-        Self::Generic {
-            err: value.to_string(),
-        }
+        Self::generic(value.to_string())
     }
 }
+
+#[macro_export]
+macro_rules! ensure {
```

Comment:
Good idea!

NIT: call it `ensure_sdk!` to avoid confusion with `anyhow::ensure!`

Reply:
Done [4edbbbe](https://github.com/breez/breez-sdk/pull/469/commits/4edbbbee20fdca35aebe326f3048148e0f9c7b48)